### DISCUSSION TITLE
Do you know about cachix/install-nix-action's `skip_adding_nixpkgs_channel`?

### DISCUSSION THREAD

**Discussion by bbigras** [lovesegfault/nix-config#114] [General]
> skip_adding_nixpkgs_channel: set to true to skip adding nixpkgs-unstable channel (and save ~5s for each job build)

https://github.com/cachix/install-nix-action#options-with-

**Comment by lovesegfault** [comment]
I didn't, but now I'm using it as of https://github.com/lovesegfault/nix-config/commit/4a2074724afa2d50c3754236e9b8555296809367

Thanks!

**Comment by bbigras** [comment]
:smile: 

Also maybe check [nix-build-uncached](https://github.com/Mic92/nix-build-uncached), I'm using it as:
`nix run '((import (import ./nix/sources.nix).nixpkgs {}).nix-build-uncached)' -c nix-build-uncached`

Maybe that will help you forgive me for stealing your config. :laughing:  Mine is at https://github.com/bbigras/nix-config.

**Comment by lovesegfault** [comment]
> Maybe that will help you forgive me for stealing your config. laughing Mine is at https://github.com/bbigras/nix-config.

Nothing to forgive! I was really happy to see someone else making use of my work! Super cool :)

**Comment by lovesegfault** [comment]
Done in 5c2c218d2cbd9a30fd3e4a51ae6d383cb762ab4e

Now using `nix-build-uncached`

**Comment by bbigras** [comment]
About `nix-build-uncached`, you might want to check https://github.com/Mic92/nix-build-uncached/issues/22 and https://github.com/cachix/cachix/issues/338.

TLDR: since `nix-build-uncached` doesn't rebuild and repush everything, some important paths can be garbage collected. The cachix dev seems open to change the gc algo.

**Comment by lovesegfault** [comment]
Aha! that explains some weirdness I think I've been seeing!

**Comment by bbigras** [comment]
yeah, I have been wondering for a while why my CI builds took so long.

I was even considering pushing to s3 instead.

What made me realize there was a problem is that I saw that the nix template repo started to also build everything when updating niv (see https://github.com/nix-dot-dev/getting-started-nix-template/blob/master/.github/workflows/update-niv.yml). I thought it was a great idea since even if I got the email in the morning about 'niv update' I still had to wait hours for it to build while being awake. With the change, merging the PR should results in a pretty quick build.

But it wasn't. It took hours too.

On an unrelated note, I saw that you seem to have a lot of audio stuff in your config, so I was wondering if you saw that pipewire PR: https://github.com/NixOS/nixpkgs/pull/93725

**Comment by bbigras** [comment]
![giphy](https://user-images.githubusercontent.com/24027/94480341-b3b37600-01c5-11eb-8530-8266a4140a0e.gif)

One more thing, it seems `skip_adding_nixpkgs_channel` is obsolete now. see https://github.com/cachix/install-nix-action/releases/tag/v11

**Comment by lovesegfault** [comment]
Looks like that has been deployed on cachix! I cleaned up the `skip_adding_nixpkgs_channel` from `ci.nix`

I added some cool stuff with `pre-commit-hooks.nix` recently that you may want to take a look at, by the way.

Regarding `pipewire` I'm only vaguely familiar with the project, what's the deal with it?

**Comment by bbigras** [comment]
> Looks like that has been deployed on cachix!

Yep, it seems fine now :smile: 

> I added some cool stuff with pre-commit-hooks.nix recently that you may want to take a look at, by the way.

Yep I stole that from your config about 7 hours ago. I checked your config today to see if you were still using `nix-build-uncached` and, while there, checked what new things you had.

I knew about `pre-commit-hooks.nix` since I saw https://github.com/nix-dot-dev/getting-started-nix-template but I had no idea how to use it with nixus. Thanks a lot for that :smile: 

> Regarding pipewire I'm only vaguely familiar with the project, what's the deal with it?

I don't know much about pipewire to be honest. I first heard about it since it seems useful for [WebRTC screen sharing on Wayland](https://wiki.archlinux.org/index.php/PipeWire#WebRTC_screen_sharing).

The website mentions "Capture and playback of audio and video with minimal latency." and "Real-time Multimedia processing on audio and video." so I thought maybe you could be interested.

It's also often advertised as pulseaudio's successor.

And it's new (appeal to novelty fallacy :laughing:)

**Comment by lovesegfault** [comment]
> I knew about `pre-commit-hooks.nix` since I saw https://github.com/nix-dot-dev/getting-started-nix-template but I had no idea how to use it with nixus. Thanks a lot for that :smile:

Nice! You may also be interested in the changes I made to the menu applications in e1541ca. I think it's a bit less weird now.

> I don't know much about pipewire to be honest. I first heard about it since it seems useful for [WebRTC screen sharing on Wayland](https://wiki.archlinux.org/index.php/PipeWire#WebRTC_screen_sharing).
> 
> The website mentions "Capture and playback of audio and video with minimal latency." and "Real-time Multimedia processing on audio and video." so I thought maybe you could be interested.
> 
> It's also often advertised as pulseaudio's successor.
> 
> And it's new (appeal to novelty fallacy :laughing:)

Oh, that does seem interesting! I'll be looking forward to seeing it land in `nixos-unstable` for me to test. Not being able to screenshare is a major pain point in Sway for me as my company uses Zoom. I've been patiently waiting to see that change for a while now.

On a separate note, my last config ambition is to get ready-to-go system images. I'd like to if (when) I get a laptop to replace `foucault` to simply do `nix-build -A system.foucault && ./result /dev/sdb`. `result` will be a script that does `sudo dd if=/nix/store/...-foucault.img of=$1 bs=1M` basically.

**Comment by bbigras** [comment]
> Oh, that does seem interesting! I'll be looking forward to seeing it land in nixos-unstable for me to test.

I think that pipewire is already in `nixos-unstable`. The new PR might be to improve stuff and maybe to make it easier to use as a drop-in pulseaudio replacement.

See: https://github.com/NixOS/nixpkgs/pull/84233

I'll reply to the rest later. :smile:

**Comment by bbigras** [comment]
> On a separate note, my last config ambition is to get ready-to-go system images. I'd like to if (when) I get a laptop to replace foucault to simply do nix-build -A system.foucault && ./result /dev/sdb. result will be a script that does sudo dd if=/nix/store/...-foucault.img of=$1 bs=1M basically.

Since you use `/dev/sdb`, does that mean that it could set up the partitions as well?

Maybe it could be nice for disaster recovery too. On my Laptop I need to make a custom ISO with the broadcom driver for the Wi-Fi to work. I'm wondering if it would be a good idea to have a `laptop.img` file always ready and if my hard drive dies, I could boot any Linux, create my partitions, run `dd` and I would have a recent enough NixOS without needing the internet. And it should be pretty fast too. I would still need to setup systemd-boot.

I wonder if it would be worth it, it's not like the internet is often down and hardware failure is pretty rare that I don't think I would get into trouble at work if it took an hour or 2 to set up my machine.

EDIT: I wonder how your image thing works with disk encryption.

**Comment by bbigras** [comment]
Btw, unstable now has [swappy](https://github.com/jtheoof/swappy) "A Wayland native snapshot editing tool, inspired by Snappy on macOS", it's pretty basic but I use those annotation tools, like flameshot when not using wayland, a lot.

**Comment by lovesegfault** [comment]
> > On a separate note, my last config ambition is to get ready-to-go system images. I'd like to if (when) I get a laptop to replace foucault to simply do nix-build -A system.foucault && ./result /dev/sdb. result will be a script that does sudo dd if=/nix/store/...-foucault.img of=$1 bs=1M basically.
> 
> Since you use `/dev/sdb`, does that mean that it could set up the partitions as well?
> 
> Maybe it could be nice for disaster recovery too. On my Laptop I need to make a custom ISO with the broadcom driver for the Wi-Fi to work. I'm wondering if it would be a good idea to have a `laptop.img` file always ready and if my hard drive dies, I could boot any Linux, create my partitions, run `dd` and I would have a recent enough NixOS without needing the internet. And it should be pretty fast too. I would still need to setup systemd-boot.
> 
> I wonder if it would be worth it, it's not like the internet is often down and hardware failure is pretty rare that I don't think I would get into trouble at work if it took an hour or 2 to set up my machine.
> 
> EDIT: I wonder how your image thing works with disk encryption.

Yeah, dealing with encryption is a challenge. Really one of the main issues has to do with partitioning, which I'm not sure how to solve yet.

**Comment by lovesegfault** [comment]
> Btw, unstable now has [swappy](https://github.com/jtheoof/swappy) "A Wayland native snapshot editing tool, inspired by Snappy on macOS", it's pretty basic but I use those annotation tools, like flameshot when not using wayland, a lot.

That's really cool! It needs a gtk wrapper though, right now clicking the colorchooser causes it to segfault :D

**Comment by bbigras** [comment]
Just a heads-up about https://serokell.io/blog/deploy-rs. It's like nixus but it works with flakes.

I started playing with it at https://github.com/bbigras/nix-config/tree/flakes. Right now nix-build-uncached doesn't seem to work in my case and I can't figure out yet how to conditionally include my secrets as a submodules. I think files need to be in the git index to work with flakes.

**Comment by lovesegfault** [comment]
@bbigras Hope you don't mind, I converted this to a discussion.

Have you been using deploy-rs + flakes now? How is it going?

**Comment by bbigras** [comment]
1. I don't know if you saw my message the other day but I got some `unexpected end-of-file` errors like this:

```
$ deploy .#desktop -- --impure
 INFO  deploy > Running checks for flake in .
warning: Git tree '/home/bbigras/nix-config' is dirty
warning: unknown flake output 'deploy'
error: --- EndOfFile -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- nix
unexpected end-of-file
 ERROR deploy > Failed to check deployment: Nix checking command resulted in a bad exit code: Some(1)
```

I think I just have to run the command again and it's usually fine. It could suck when using Github actions.

---

2. There's something fishy with the update PRs. Maybe it's the same problem I had the other day.

![image](https://user-images.githubusercontent.com/24027/101973260-b05d5180-3c2e-11eb-8cb7-78369a93b15f.png)

The first one is when I merged it. It makes no sense that the merge build took more time.

---

3. I had some performances issues while running `deploy` but I'm suspecting something is wrong with my firefox profile and maybe paper-wm.

Using a fresh firefox profile seemed to help.

In the paper-wm case, I get some warning messages since a recent update. Maybe it has some problems.

I'm often doing a lot of stuff like building rust code at the same time so maybe it's not really a problem.

**Reply by lovesegfault** [reply]
Oh, interesting, I think this is sufficient for me to wait a little bit longer to switch over.

I use `nix copy` a lot for my workflow and `nixUnstable` has had issues with that too recently.

**Comment by bbigras** [comment]
Another thing is that it seems to download/copy a lot of GB, but I didn't see how much it was downloading before using flakes. It did take a while to deploy with nixus so maybe it was doing as much then.

**Reply by lovesegfault** [reply]
That's pretty interesting, I wonder why!
